File: apps/api/src/istari_service/product_types.py
"""Persistence-independent managed-product types and lifecycle values."""

from __future__ import annotations

from collections.abc import AsyncIterator
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from uuid import UUID


class PackageStatus(StrEnum):
    DRAFT = "DRAFT"
    REVIEW_READY = "REVIEW_READY"
    MANAGER_APPROVED = "MANAGER_APPROVED"
    DISSEMINATED = "DISSEMINATED"
    REPLACED = "REPLACED"
    WITHDRAWN = "WITHDRAWN"


class ArtefactKind(StrEnum):
    MANAGED_FILE = "MANAGED_FILE"
    EXTERNAL_LINK = "EXTERNAL_LINK"


class ArtefactLifecycle(StrEnum):
    PENDING_UPLOAD = "PENDING_UPLOAD"
    QUARANTINED = "QUARANTINED"
    CLEAN = "CLEAN"
    RELEASED = "RELEASED"
    FAILED = "FAILED"
    REPLACED = "REPLACED"
    WITHDRAWN = "WITHDRAWN"
    EXPIRED = "EXPIRED"


class ScanResult(StrEnum):
    CLEAN = "CLEAN"
    FAILED = "FAILED"
    UNKNOWN = "UNKNOWN"
    TIMED_OUT = "TIMED_OUT"


class AccessKind(StrEnum):
    DOWNLOAD = "DOWNLOAD"
    REDIRECT = "REDIRECT"


class AccessOutcome(StrEnum):
    ALLOWED = "ALLOWED"
    DENIED = "DENIED"
    UNAVAILABLE = "UNAVAILABLE"


@dataclass(frozen=True, slots=True)
class UploadGrant:
    """Opaque single-purpose target, never a public object URL."""

    object_key: str
    token: str
    expires_at: datetime


@dataclass(frozen=True, slots=True)
class StoredObject:
    size_bytes: int
    media_type: str
    checksum: str


@dataclass(frozen=True, slots=True)
class ScanDecision:
    result: ScanResult
    scanner: str
    scanner_version: str
    reason_code: str | None = None


@dataclass(frozen=True, slots=True)
class DownloadStream:
    chunks: AsyncIterator[bytes]
    media_type: str
    filename: str


@dataclass(frozen=True, slots=True)
class AccessAuditRecord:
    request_id: UUID | None
    package_id: UUID | None
    artefact_id: UUID | None
    target_reference: UUID
    actor_id: UUID
    kind: AccessKind
    outcome: AccessOutcome
    reason_code: str
    correlation_id: str | None = None


@dataclass(frozen=True, slots=True)
class ProductRequestRecord:
    id: UUID
    requester_id: UUID
    status: str
    assigned_team: str | None
    assigned_specialist_id: UUID | None
    version: int
    assigned_team_id: UUID | None = None
    participant_ids: frozenset[UUID] = frozenset()


@dataclass(frozen=True, slots=True)
class PackageRecord:
    id: UUID
    request_id: UUID
    author_user_id: UUID
    status: PackageStatus
    package_checksum: str | None
    version: int
    package_version: int


@dataclass(frozen=True, slots=True)
class ArtefactRecord:
    id: UUID
    package_id: UUID
    kind: ArtefactKind
    lifecycle: ArtefactLifecycle
    filename: str | None
    media_type: str | None
    size_bytes: int | None
    checksum: str | None
    quarantine_key: str | None
    released_key: str | None
    destination_url: str | None = None
    destination_domain: str | None = None
    expires_at: datetime | None = None


@dataclass(frozen=True, slots=True)
class UploadIntentRecord:
    id: UUID
    artefact_id: UUID
    object_key: str
    expected_size_bytes: int
    expected_media_type: str
    expected_checksum: str
    expires_at: datetime
    uploaded_at: datetime | None
    consumed_at: datetime | None


@dataclass(frozen=True, slots=True)
class ReleaseAccessRecord:
    request_id: UUID
    package_id: UUID
    artefact: ArtefactRecord

File: apps/web/src/lib/api/productTypes.ts
export type ProductPackageStatus =
  "DRAFT" | "REVIEW_READY" | "MANAGER_APPROVED" | "DISSEMINATED" | "REPLACED" | "WITHDRAWN";

export type ProductArtefactLifecycle =
  | "PENDING_UPLOAD"
  | "QUARANTINED"
  | "CLEAN"
  | "RELEASED"
  | "FAILED"
  | "REPLACED"
  | "WITHDRAWN"
  | "EXPIRED";

export interface ProductArtefact {
  id: string;
  packageId: string;
  position: number;
  kind: "MANAGED_FILE" | "EXTERNAL_LINK";
  lifecycle: ProductArtefactLifecycle;
  label: string;
  filename: string | null;
  mediaType: string | null;
  sizeBytes: number | null;
  sha256: string | null;
  destinationDomain: string | null;
  reviewDestinationUrl?: string | null;
  reviewUrl?: string | null;
  expiresAt: string | null;
  scanResult: "CLEAN" | "FAILED" | "UNKNOWN" | "TIMED_OUT" | null;
  scanReason: string | null;
  releasedAt: string | null;
  version: number;
}

export interface ProductPackage {
  id: string;
  requestId: string;
  requestReference: string;
  requestTitle: string;
  requestStatus: string;
  packageVersion: number;
  policyVersion: number;
  status: ProductPackageStatus;
  coveringNote: string | null;
  packageChecksum: string | null;
  version: number;
  authorDisplayName: string;
  managerApprovedBy: string | null;
  managerApprovedAt: string | null;
  disseminatedBy: string | null;
  disseminatedAt: string | null;
  withdrawalReason: string | null;
  artefacts: ProductArtefact[];
}

export interface UploadIntent {
  id: string;
  objectKey: string;
  uploadToken: string;
  expiresAt: string;
}

export interface ManagedArtefactResult {
  package: ProductPackage;
  artefact: ProductArtefact;
  uploadIntent: UploadIntent;
}

export interface ProductRelease {
  packageId: string;
  requestId: string;
  packageVersion: number;
  status: "DISSEMINATED" | "WITHDRAWN" | "REPLACED";
  releasedAt: string;
  releasedBy: string;
  coveringNote: string;
  acceptedAt: string | null;
  artefacts: ProductArtefact[];
}

export interface UploadContentReceipt {
  intentId: string;
  sizeBytes: number;
  sha256: string;
  uploadedAt: string;
  packageVersion: number;
}

export interface ManagedArtefactInput {
  expectedVersion: number;
  label: string;
  filename: string;
  mediaType: string;
  sizeBytes: number;
  sha256: string;
  idempotencyKey: string;
}

export interface ExternalLinkInput {
  expectedVersion: number;
  label: string;
  url: string;
  expiresAt?: string;
  idempotencyKey: string;
}
